Make go generate directives work on more setups.

shurcooL opened the pull request:
Add support for running go generate successfully on OS X.

Add support for running go generate with multiple GOPATH workspaces.

slimsag commented:
LGTM

shurcooL commented:
When this is done being reviewed and ready to merge, I'm happy to move `gopathexec` under `sourcegraph` namespace if desired.

sqs commented on README.md:
````diff
@@ -37,12 +38,35 @@ regenerate Go code after you change the `sourcegraph.proto` file.
+#### OS X-specific
+
+```
+brew install gnu-sed --with-default-names
````
Do other OS X tools/scripts rely on non-GNU sed behavior? This could mess up people's computers if so.

shurcooL replied:
Really hard to know.

I agree this is a potentially invasive change that not everyone might want to do (if they depend on existing OS X-specific `sed` for some other things). The only alternatives I can think of:

1. Use a namespaced version of sed somehow. So if you remove the `--with-default-names` option, it uses `gsed` binary name. So we can remove the `--with-default-names` flag, and add Linux section that does `alias gsed='sed'` or so.

2. Don't use `sed` since it's not portable. Perhaps write a simple (portable) Go tool that does what we want (does not need to be general like sed, just enough to perform the needed task, which is simple).

sqs replied:
what about making the go:generate line call `sh -c` and use simple shell scripting to switch between OS X and Linux? (checking `uname` or something.) Or something like that.

shurcooL replied:
That should work (assuming `sh` exists, should be safe on OS X and Linux, but not Windows), albeit being somewhat underhanded.

What's the advantage over doing `alias gsed='sed'` on Linux? Is it so that the Linux namespace doesn't need to be polluted with an extra command/alias?

sqs commented:
lgtm except 1 comment

shurcooL commented:
Addressed the comment, PTAL.

After too much time spent considering/looking for a portable way of using sed/perl/awk or similar, I decided for the simple use of sed, it's nicer just to replace it with a 45 line Go program (which would even work on Windows).

sqs commented:
This'll work, but I just was looking and we only use sed here (and nowhere else). We might be able to eliminate sed entirely by putting `pbtypes.Timestamp` and `pbtypes.Void` in the same `.proto` file.

shurcooL commented:
Since this PR works (it resolves the original issue), if it LGTY, what do you think about merging it as is, and removing `gen/gorename.go` in an optional followup?

sqs commented:
sgtm

shurcooL commented:
I believe everything should work without problems (make note of the updated steps in the README) from now on. Please report any issues if you do encounter them.

shurcooL commented:
I've created a similar change for srclib at https://github.com/sourcegraph/srclib/pull/177.
